File: android/app/src/main/kotlin/com/androidbridge/core/SyncJournal.kt
package com.androidbridge.core

import com.androidbridge.protocol.ConflictOutcome
import com.androidbridge.protocol.MAX_FRAME_PAYLOAD_BYTES
import com.androidbridge.protocol.SyncOperation
import com.androidbridge.protocol.SyncOperationKind
import com.androidbridge.protocol.SyncModelCodec
import com.androidbridge.protocol.TransferChunk
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.security.MessageDigest
import java.util.Base64
import java.util.UUID
import kotlinx.serialization.Serializable

object ContentHash {
    fun sha256(bytes: ByteArray): String = MessageDigest.getInstance("SHA-256")
        .digest(bytes)
        .joinToString("") { "%02x".format(it) }
}

enum class SyncJournalError {
    CURSOR_REGRESSION,
    CURSOR_BEYOND_HIGH_WATER,
    STALE_CURSOR,
    SEQUENCE_GAP,
    OPERATION_ID_COLLISION,
    DIGEST_MISMATCH,
    INVALID_CHUNK,
    INCOMPLETE_TRANSFER,
    INVALID_OPERATION,
}

class SyncJournalException(val error: SyncJournalError) : Exception(error.name)

enum class IncomingDisposition { APPLY, DUPLICATE, GAP }

@Serializable
private data class ReceivedOperation(val actorId: String, val sequence: Long)

@Serializable
private data class JournalState(
    val actorId: String,
    val highWater: Long = 0,
    val acknowledgedThrough: Long = 0,
    val operations: List<SyncOperation> = emptyList(),
    val localOperationIds: Set<String> = emptySet(),
    val receivedCursors: Map<String, Long> = emptyMap(),
    val receivedOperations: Map<String, ReceivedOperation> = emptyMap(),
)

class DurableSyncJournal(private val root: File, private val actorId: String) {
    private val stateFile = File(root, "journal.json")
    private val blobDirectory = File(root, "blobs")
    private var state: JournalState

    init {
        require(actorId.isNotEmpty())
        if (!blobDirectory.mkdirs() && !blobDirectory.isDirectory) error("cannot create sync journal")
        state = if (stateFile.exists()) {
            SyncModelCodec.decode(stateFile.readText())
        } else {
            JournalState(actorId)
        }
        require(state.actorId == actorId) { "journal actor mismatch" }
        validateLoadedState()
    }

    val highWater: Long @Synchronized get() = state.highWater
    val acknowledgedThrough: Long @Synchronized get() = state.acknowledgedThrough

    @Synchronized
    fun enqueue(
        operationId: String,
        kind: SyncOperationKind,
        target: String,
        content: ByteArray?,
        baseDigest: String? = null,
        messageType: String? = null,
        mediaType: String? = null,
    ): SyncOperation {
        validateMutation(operationId, kind, target, content, messageType)
        if (operationId in state.localOperationIds) throw SyncJournalException(SyncJournalError.OPERATION_ID_COLLISION)
        val digest = content?.let {
            val hash = ContentHash.sha256(it)
            storeBlob(hash, it)
            hash
        }
        val operation = SyncOperation(
            operationId = operationId,
            actorId = actorId,
            sequence = state.highWater + 1,
            kind = kind,
            target = target,
            messageType = messageType,
            baseDigest = baseDigest,
            resultDigest = digest,
            blobDigest = digest,
            byteCount = content?.size?.toLong() ?: 0,
            mediaType = mediaType,
        )
        commit(state.copy(
            highWater = operation.sequence,
            operations = state.operations + operation,
            localOperationIds = state.localOperationIds + operationId,
        ))
        return operation
    }

    @Synchronized
    fun pending(): List<SyncOperation> = state.operations

    @Synchronized
    fun pending(afterSequence: Long): List<SyncOperation> {
        if (afterSequence < state.acknowledgedThrough) throw SyncJournalException(SyncJournalError.STALE_CURSOR)
        if (afterSequence > state.highWater) throw SyncJournalException(SyncJournalError.CURSOR_BEYOND_HIGH_WATER)
        return state.operations.filter { it.sequence > afterSequence }
    }

    @Synchronized
    fun acknowledge(throughSequence: Long) {
        if (throughSequence < state.acknowledgedThrough) throw SyncJournalException(SyncJournalError.CURSOR_REGRESSION)
        if (throughSequence > state.highWater) throw SyncJournalException(SyncJournalError.CURSOR_BEYOND_HIGH_WATER)
        if (throughSequence == state.acknowledgedThrough) return
        commit(state.copy(
            acknowledgedThrough = throughSequence,
            operations = state.operations.filter { it.sequence > throughSequence },
        ))
        pruneUnreferencedBlobs()
    }

    @Synchronized
    fun incomingDisposition(operation: SyncOperation): IncomingDisposition {
        val seen = state.receivedOperations[operation.operationId]
        if (seen != null) {
            if (seen.actorId != operation.actorId || seen.sequence != operation.sequence) {
                throw SyncJournalException(SyncJournalError.OPERATION_ID_COLLISION)
            }
            return IncomingDisposition.DUPLICATE
        }
        val cursor = state.receivedCursors[operation.actorId] ?: 0
        if (operation.sequence <= cursor) throw SyncJournalException(SyncJournalError.STALE_CURSOR)
        return if (operation.sequence == cursor + 1) IncomingDisposition.APPLY else IncomingDisposition.GAP
    }

    @Synchronized
    fun recordApplied(operation: SyncOperation): Boolean = when (incomingDisposition(operation)) {
        IncomingDisposition.DUPLICATE -> false
        IncomingDisposition.GAP -> throw SyncJournalException(SyncJournalError.SEQUENCE_GAP)
        IncomingDisposition.APPLY -> {
            commit(state.copy(
                receivedCursors = state.receivedCursors + (operation.actorId to operation.sequence),
                receivedOperations = state.receivedOperations +
                    (operation.operationId to ReceivedOperation(operation.actorId, operation.sequence)),
            ))
            true
        }
    }

    @Synchronized
    fun receivedThrough(actorId: String): Long = state.receivedCursors[actorId] ?: 0

    @Synchronized
    fun readBlob(digest: String): ByteArray {
        if (!digest.matches(Regex("[0-9a-f]{64}"))) throw SyncJournalException(SyncJournalError.DIGEST_MISMATCH)
        val bytes = File(blobDirectory, digest).readBytes()
        if (ContentHash.sha256(bytes) != digest) throw SyncJournalException(SyncJournalError.DIGEST_MISMATCH)
        return bytes
    }

    private fun validateMutation(
        operationId: String,
        kind: SyncOperationKind,
        target: String,
        content: ByteArray?,
        messageType: String?,
    ) {
        val invalid = operationId.isEmpty() || target.isEmpty() ||
            (kind == SyncOperationKind.TOMBSTONE && content != null) ||
            (kind != SyncOperationKind.TOMBSTONE && content == null) ||
            (kind == SyncOperationKind.MESSAGE && messageType == null)
        if (invalid) throw SyncJournalException(SyncJournalError.INVALID_OPERATION)
    }

    private fun validateLoadedState() {
        require(state.acknowledgedThrough in 0..state.highWater)
        require(state.operations.zipWithNext().all { (a, b) -> b.sequence == a.sequence + 1 })
        require(state.operations.isEmpty() || state.operations.first().sequence == state.acknowledgedThrough + 1)
        require(state.operations.isEmpty() || state.operations.last().sequence == state.highWater)
        require(state.operations.all { it.actorId == actorId && it.sequence > state.acknowledgedThrough })
        state.operations.mapNotNull { it.blobDigest }.forEach { readBlob(it) }
    }

    private fun storeBlob(digest: String, bytes: ByteArray) {
        val destination = File(blobDirectory, digest)
        if (destination.exists()) {
            if (!destination.readBytes().contentEquals(bytes)) throw SyncJournalException(SyncJournalError.DIGEST_MISMATCH)
            return
        }
        atomicWrite(destination, bytes)
    }

    private fun commit(next: JournalState) {
        atomicWrite(stateFile, SyncModelCodec.encode(next).encodeToByteArray())
        state = next
    }

    private fun pruneUnreferencedBlobs() {
        val retained = state.operations.mapNotNull { it.blobDigest }.toSet()
        blobDirectory.listFiles().orEmpty()
            .filter { it.isFile && it.name !in retained }
            .forEach { if (!it.delete()) error("cannot remove acknowledged sync blob") }
    }

    private fun atomicWrite(destination: File, bytes: ByteArray) {
        val temporary = File(destination.parentFile, ".${destination.name}.${UUID.randomUUID()}.tmp")
        try {
            FileOutputStream(temporary).use { output ->
                output.write(bytes)
                output.fd.sync()
            }
            Files.move(
                temporary.toPath(),
                destination.toPath(),
                StandardCopyOption.ATOMIC_MOVE,
                StandardCopyOption.REPLACE_EXISTING,
            )
        } finally {
            temporary.delete()
        }
    }
}

data class NoteConflictResolution(
    val outcome: ConflictOutcome,
    val canonical: ByteArray?,
    val conflict: ByteArray?,
)

object NoteConflictResolver {
    fun resolve(current: ByteArray?, incoming: ByteArray?, baseDigest: String?): NoteConflictResolution {
        val currentDigest = current?.let(ContentHash::sha256)
        val incomingDigest = incoming?.let(ContentHash::sha256)
        if (currentDigest == incomingDigest) {
            return NoteConflictResolution(ConflictOutcome.UNCHANGED, current, null)
        }
        if (currentDigest == baseDigest) {
            val outcome = if (incoming == null) ConflictOutcome.DELETED else ConflictOutcome.APPLIED
            return NoteConflictResolution(outcome, incoming, null)
        }
        return NoteConflictResolution(ConflictOutcome.CONFLICT_PRESERVED, current, incoming)
    }
}

object SyncTransferChunker {
    fun chunk(operationId: String, data: ByteArray, chunkSize: Int): List<TransferChunk> {
        if (chunkSize !in 1..MAX_FRAME_PAYLOAD_BYTES.toInt()) throw SyncJournalException(SyncJournalError.INVALID_CHUNK)
        if (data.isEmpty()) return listOf(makeChunk(operationId, 0, 0, ByteArray(0), true))
        return data.indices.step(chunkSize).mapIndexed { index, offset ->
            val end = minOf(offset + chunkSize, data.size)
            makeChunk(operationId, index, offset.toLong(), data.copyOfRange(offset, end), end == data.size)
        }
    }

    private fun makeChunk(operationId: String, index: Int, offset: Long, bytes: ByteArray, final: Boolean) =
        TransferChunk(
            operationId = operationId,
            index = index,
            offset = offset,
            dataBase64 = Base64.getEncoder().encodeToString(bytes),
            digest = ContentHash.sha256(bytes),
            isFinal = final,
        )
}

class SyncTransferReassembler(
    private val operationId: String,
    private val expectedDigest: String,
) {
    private val output = ByteArrayOutputStream()
    private var nextIndex = 0
    private var complete = false

    fun accept(chunk: TransferChunk) {
        if (complete || chunk.operationId != operationId || chunk.index != nextIndex || chunk.offset != output.size().toLong()) {
            throw SyncJournalException(SyncJournalError.INVALID_CHUNK)
        }
        val bytes = try {
            chunk.decodedData()
        } catch (_: IllegalArgumentException) {
            throw SyncJournalException(SyncJournalError.INVALID_CHUNK)
        }
        if (bytes.size > MAX_FRAME_PAYLOAD_BYTES || ContentHash.sha256(bytes) != chunk.digest) {
            throw SyncJournalException(SyncJournalError.DIGEST_MISMATCH)
        }
        output.write(bytes)
        nextIndex++
        complete = chunk.isFinal
    }

    fun result(): ByteArray {
        if (!complete) throw SyncJournalException(SyncJournalError.INCOMPLETE_TRANSFER)
        val result = output.toByteArray()
        if (ContentHash.sha256(result) != expectedDigest) throw SyncJournalException(SyncJournalError.DIGEST_MISMATCH)
        return result
    }
}

File: mac/Sources/BridgeCore/SyncJournal.swift
import Foundation
import Crypto
import DeviceLinkProtocol

public enum ContentHash {
    public static func sha256(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
}

public enum SyncJournalError: Error, Equatable {
    case cursorRegression
    case cursorBeyondHighWater
    case staleCursor
    case sequenceGap
    case operationIdCollision
    case digestMismatch
    case invalidChunk
    case incompleteTransfer
    case invalidOperation
}

public enum IncomingDisposition: Equatable { case apply, duplicate, gap }

private struct ReceivedOperation: Codable {
    let actorId: String
    let sequence: Int64
}

private struct JournalState: Codable {
    let actorId: String
    var highWater: Int64 = 0
    var acknowledgedThrough: Int64 = 0
    var operations: [SyncOperation] = []
    var localOperationIds: Set<String> = []
    var receivedCursors: [String: Int64] = [:]
    var receivedOperations: [String: ReceivedOperation] = [:]
}

public final class DurableSyncJournal {
    private let stateURL: URL
    private let blobURL: URL
    private let actorId: String
    private let lock = NSLock()
    private let incomingApplyLock = NSLock()
    private var state: JournalState
    private let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return encoder
    }()

    public init(rootURL: URL, actorId: String) throws {
        precondition(!actorId.isEmpty)
        self.stateURL = rootURL.appendingPathComponent("journal.json")
        self.blobURL = rootURL.appendingPathComponent("blobs", isDirectory: true)
        self.actorId = actorId
        try FileManager.default.createDirectory(at: self.blobURL, withIntermediateDirectories: true)
        if FileManager.default.fileExists(atPath: self.stateURL.path) {
            self.state = try JSONDecoder().decode(JournalState.self, from: Data(contentsOf: self.stateURL))
        } else {
            self.state = JournalState(actorId: actorId)
        }
        precondition(state.actorId == actorId, "journal actor mismatch")
        try validateLoadedState()
    }

    public var highWater: Int64 { withLock { state.highWater } }
    public var acknowledgedThrough: Int64 { withLock { state.acknowledgedThrough } }

    @discardableResult
    public func enqueue(
        operationId: String,
        kind: SyncOperationKind,
        target: String,
        content: Data?,
        baseDigest: String? = nil,
        messageType: String? = nil,
        mediaType: String? = nil
    ) throws -> SyncOperation {
        try withLock {
            try validateMutation(operationId: operationId, kind: kind, target: target, content: content, messageType: messageType)
            if state.localOperationIds.contains(operationId) { throw SyncJournalError.operationIdCollision }
            let digest = try content.map { data -> String in
                let hash = ContentHash.sha256(data)
                try storeBlob(digest: hash, data: data)
                return hash
            }
            let operation = SyncOperation(
                operationId: operationId,
                actorId: actorId,
                sequence: state.highWater + 1,
                kind: kind,
                target: target,
                messageType: messageType,
                baseDigest: baseDigest,
                resultDigest: digest,
                blobDigest: digest,
                byteCount: Int64(content?.count ?? 0),
                mediaType: mediaType
            )
            var next = state
            next.highWater = operation.sequence
            next.operations.append(operation)
            next.localOperationIds.insert(operationId)
            try commit(next)
            return operation
        }
    }

    public func pending() throws -> [SyncOperation] { withLock { state.operations } }

    public func pending(after sequence: Int64) throws -> [SyncOperation] {
        try withLock {
            if sequence < state.acknowledgedThrough { throw SyncJournalError.staleCursor }
            if sequence > state.highWater { throw SyncJournalError.cursorBeyondHighWater }
            return state.operations.filter { $0.sequence > sequence }
        }
    }

    public func acknowledge(through sequence: Int64) throws {
        try withLock {
            if sequence < state.acknowledgedThrough { throw SyncJournalError.cursorRegression }
            if sequence > state.highWater { throw SyncJournalError.cursorBeyondHighWater }
            if sequence == state.acknowledgedThrough { return }
            var next = state
            next.acknowledgedThrough = sequence
            next.operations.removeAll { $0.sequence <= sequence }
            try commit(next)
            try pruneUnreferencedBlobs()
        }
    }

    public func incomingDisposition(_ operation: SyncOperation) throws -> IncomingDisposition {
        try withLock { try disposition(operation) }
    }

    @discardableResult
    public func recordApplied(_ operation: SyncOperation) throws -> Bool {
        try recordApplied(operation, durableApply: {})
    }

    @discardableResult
    public func recordApplied(_ operation: SyncOperation, durableApply: () throws -> Void) throws -> Bool {
        incomingApplyLock.lock()
        defer { incomingApplyLock.unlock() }
        switch try withLock({ try disposition(operation) }) {
        case .duplicate:
            return false
        case .gap:
            throw SyncJournalError.sequenceGap
        case .apply:
            try durableApply()
            return try withLock {
                var next = state
                next.receivedCursors[operation.actorId] = operation.sequence
                next.receivedOperations[operation.operationId] = ReceivedOperation(actorId: operation.actorId, sequence: operation.sequence)
                try commit(next)
                return true
            }
        }
    }

    public func receivedThrough(actorId: String) throws -> Int64 {
        withLock { state.receivedCursors[actorId] ?? 0 }
    }

    public func readBlob(digest: String) throws -> Data {
        try withLock { try readBlobUnlocked(digest: digest) }
    }

    private func readBlobUnlocked(digest: String) throws -> Data {
        guard digest.range(of: "^[0-9a-f]{64}$", options: .regularExpression) != nil else {
            throw SyncJournalError.digestMismatch
        }
        let data = try Data(contentsOf: blobURL.appendingPathComponent(digest))
        guard ContentHash.sha256(data) == digest else { throw SyncJournalError.digestMismatch }
        return data
    }

    private func disposition(_ operation: SyncOperation) throws -> IncomingDisposition {
        if let seen = state.receivedOperations[operation.operationId] {
            guard seen.actorId == operation.actorId, seen.sequence == operation.sequence else {
                throw SyncJournalError.operationIdCollision
            }
            return .duplicate
        }
        let cursor = state.receivedCursors[operation.actorId] ?? 0
        if operation.sequence <= cursor { throw SyncJournalError.staleCursor }
        return operation.sequence == cursor + 1 ? .apply : .gap
    }

    private func validateMutation(
        operationId: String,
        kind: SyncOperationKind,
        target: String,
        content: Data?,
        messageType: String?
    ) throws {
        let invalid = operationId.isEmpty || target.isEmpty ||
            (kind == .tombstone && content != nil) ||
            (kind != .tombstone && content == nil) ||
            (kind == .message && messageType == nil)
        if invalid { throw SyncJournalError.invalidOperation }
    }

    private func validateLoadedState() throws {
        precondition(state.acknowledgedThrough >= 0 && state.acknowledgedThrough <= state.highWater)
        precondition(zip(state.operations, state.operations.dropFirst()).allSatisfy { $0.1.sequence == $0.0.sequence + 1 })
        precondition(state.operations.isEmpty || state.operations.first?.sequence == state.acknowledgedThrough + 1)
        precondition(state.operations.isEmpty || state.operations.last?.sequence == state.highWater)
        precondition(state.operations.allSatisfy { $0.actorId == actorId && $0.sequence > state.acknowledgedThrough })
        for digest in state.operations.compactMap(\.blobDigest) { _ = try readBlobUnlocked(digest: digest) }
    }

    private func storeBlob(digest: String, data: Data) throws {
        let destination = blobURL.appendingPathComponent(digest)
        if FileManager.default.fileExists(atPath: destination.path) {
            guard try Data(contentsOf: destination) == data else { throw SyncJournalError.digestMismatch }
            return
        }
        try data.write(to: destination, options: .atomic)
    }

    private func commit(_ next: JournalState) throws {
        try encoder.encode(next).write(to: stateURL, options: .atomic)
        state = next
    }

    private func pruneUnreferencedBlobs() throws {
        let retained = Set(state.operations.compactMap(\.blobDigest))
        let blobs = try FileManager.default.contentsOfDirectory(at: blobURL, includingPropertiesForKeys: nil)
        for blob in blobs where !retained.contains(blob.lastPathComponent) {
            try FileManager.default.removeItem(at: blob)
        }
    }

    private func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try body()
    }
}

public struct NoteConflictResolution: Equatable {
    public let outcome: ConflictOutcome
    public let canonical: Data?
    public let conflict: Data?
}

public enum NoteConflictResolver {
    public static func resolve(current: Data?, incoming: Data?, baseDigest: String?) -> NoteConflictResolution {
        let currentDigest = current.map(ContentHash.sha256)
        let incomingDigest = incoming.map(ContentHash.sha256)
        if currentDigest == incomingDigest {
            return NoteConflictResolution(outcome: .unchanged, canonical: current, conflict: nil)
        }
        if currentDigest == baseDigest {
            let outcome: ConflictOutcome = incoming == nil ? .deleted : .applied
            return NoteConflictResolution(outcome: outcome, canonical: incoming, conflict: nil)
        }
        return NoteConflictResolution(outcome: .conflictPreserved, canonical: current, conflict: incoming)
    }
}

public enum SyncTransferChunker {
    public static func chunk(operationId: String, data: Data, chunkSize: Int) throws -> [TransferChunk] {
        guard (1...ProtocolConstants.maxFramePayloadBytes).contains(chunkSize) else {
            throw SyncJournalError.invalidChunk
        }
        if data.isEmpty { return [makeChunk(operationId: operationId, index: 0, offset: 0, data: Data(), final: true)] }
        var chunks: [TransferChunk] = []
        var offset = 0
        while offset < data.count {
            let end = min(offset + chunkSize, data.count)
            let bytes = data.subdata(in: offset..<end)
            chunks.append(makeChunk(operationId: operationId, index: chunks.count, offset: Int64(offset), data: bytes, final: end == data.count))
            offset = end
        }
        return chunks
    }

    private static func makeChunk(operationId: String, index: Int, offset: Int64, data: Data, final: Bool) -> TransferChunk {
        TransferChunk(
            operationId: operationId,
            index: index,
            offset: offset,
            dataBase64: data.base64EncodedString(),
            digest: ContentHash.sha256(data),
            isFinal: final
        )
    }
}

public final class SyncTransferReassembler {
    private let operationId: String
    private let expectedDigest: String
    private var data = Data()
    private var nextIndex = 0
    private var complete = false

    public init(operationId: String, expectedDigest: String) {
        self.operationId = operationId
        self.expectedDigest = expectedDigest
    }

    public func accept(_ chunk: TransferChunk) throws {
        guard !complete,
              chunk.operationId == operationId,
              chunk.index == nextIndex,
              chunk.offset == Int64(data.count) else {
            throw SyncJournalError.invalidChunk
        }
        let bytes: Data
        do { bytes = try chunk.decodedData() } catch { throw SyncJournalError.invalidChunk }
        guard bytes.count <= ProtocolConstants.maxFramePayloadBytes,
              ContentHash.sha256(bytes) == chunk.digest else {
            throw SyncJournalError.digestMismatch
        }
        data.append(bytes)
        nextIndex += 1
        complete = chunk.isFinal
    }

    public func result() throws -> Data {
        guard complete else { throw SyncJournalError.incompleteTransfer }
        guard ContentHash.sha256(data) == expectedDigest else { throw SyncJournalError.digestMismatch }
        return data
    }
}
